### PR TITLE
core: implement a method to get secure time

### PR DESCRIPTION
**Pull request by weizhaojiang**
Add new command (STATS_CMD_SECTIME_STATS) in the stats PTA. The command returns the tee and ree time for debug purposes. https://github.com/OP-TEE/optee_os/issues/6304

Tested-by: Weizhao Jiang <weizhaoj@amazon.com>

### CODE REVIEW

File: core/pta/stats.c [hunk 1]
```diff
@@ -175,6 +177,29 @@ static TEE_Result get_user_ta_stats(uint32_t type,
 	return res;
 }
 
+static TEE_Result get_securetime_stats(uint32_t type,
+				       TEE_Param p[TEE_NUM_PARAMS])
+{
+	TEE_Time tee_time;
+	TEE_Time ree_time;
```
Comment (etienne-lms) [inline core/pta/stats.c:184]: `= { }` for both.

File: core/pta/stats.c [hunk 2]
```diff
@@ -37,6 +38,7 @@
  * uint32_t    Biggest byte size which allocation succeeded
  */
 #define STATS_CMD_TA_STATS		3
+#define STATS_CMD_SECTIME_STATS		4
```
Comment (etienne-lms) [inline core/pta/stats.c:41]: I suggest to rename the command to `STATS_CMD_SYSTEM_TIME`
Please add an inline description of the ABI, e.g.:
```c
/*
 * STATS_CMD_SYSTEM_TIME - Get system times
 * 
 * [out]    value[0].a        REE time as seen by OP-TEE in seconds
 * [out]    value[0].b        REE time as seen by OP-TEE, milliseconds part
 * [out]    value[1].a        TEE system time in seconds
 * [out]    value[1].b        TEE system time, milliseconds part
 */
#define STATS_CMD_SECTIME_STATS		4
```

File: core/pta/stats.c [hunk 3]
```diff
@@ -11,6 +11,7 @@
 #include <string.h>
 #include <string_ext.h>
 #include <malloc.h>
+#include <kernel/tee_time.h>
```
Comment (etienne-lms) [inline core/pta/stats.c:14]: Shoudl follow the alphabetical order.
Bt the way, could you also the other `#include` directives?
```c
#include <compiler.h>
#include <kernel/pseudo_ta.h>
#include <kernel/tee_time.h>
#include <malloc.h>
#include <mm/tee_mm.h>
#include <mm/tee_pager.h>
#include <stdio.h>
#include <string.h>
#include <string_ext.h>
#include <trace.h>
```

**Comment by etienne-lms** [comment]
Would you mind creating a P-R in optee_test also to add an `xtest --stats --time` command?

File: core/pta/stats.c (hunk 2 above)
Comment (weizhaojiang) [reply core/pta/stats.c:41]: good points, thanks.

File: core/pta/stats.c (hunk 3 above)
Comment (weizhaojiang) [reply core/pta/stats.c:14]: will do.

**Comment by jforissier** [comment]
Why not use a normal TA for this (`TEE_GetSystemTime()` and `TEE_GetREETime()`)?

**Comment by weizhaojiang** [comment]
> Why not use a normal TA for this (`TEE_GetSystemTime()` and `TEE_GetREETime()`)?

Thanks for your comment. We would like to add this cmd in ``stats`` PTA rather than a user TA because:
1. The cmd is to query secure time which is implemented in OS kernel. ``TEE_GetSystemTime()`` and ``TEE_GetREETime()`` essentially call into ``tee_time_get_sys_time()`` and ``tee_time_get_ree_time``.
2. We hope to test TEE secure without loading a user TA. This can simplifying our test procedure design.
3. If we add the cmd into a specific user TA, that means only that TA's client can get TEE time for testing. In our product, the TA is on-demand built into system, so it might not be available in that product. The ``stats`` PTA looks a common place to add and it will make the cmd available for all products.

File: core/pta/stats.c [hunk 4]
```diff
@@ -175,6 +185,29 @@ static TEE_Result get_user_ta_stats(uint32_t type,
 	return res;
 }
 
+static TEE_Result get_system_time(uint32_t type,
+				  TEE_Param p[TEE_NUM_PARAMS])
+{
+	TEE_Time ree_time = {0, 0};
+	TEE_Time tee_time = {0, 0};
```
[the remark this answers is not in the document]
Comment (weizhaojiang) [reply core/pta/stats.c:192]: sure, will do. Thanks for suggestion.

File: core/pta/stats.c [hunk 5]
```diff
@@ -37,6 +38,15 @@
  * uint32_t    Biggest byte size which allocation succeeded
  */
 #define STATS_CMD_TA_STATS		3
+/*
```
[the remark this answers is not in the document]
Comment (weizhaojiang) [reply core/pta/stats.c:41]: will do.

File: core/pta/stats.c [hunk 6]
```diff
@@ -36,7 +37,17 @@
...
+ * STATS_CMD_SYSTEM_TIME - Get system times
+ *
+ * [out]    value[0].a        REE time as seen by OP-TEE in seconds
+ * [out]    value[0].b        REE time as seen by OP-TEE, milliseconds part
+ * [out]    value[1].a        TEE system time in seconds
+ * [out]    value[1].b        TEE system time, milliseconds part
+ */
+#define STATS_CMD_SYSTEM_TIME		4
```
Comment (jforissier) [inline core/pta/stats.c:50]: Since it's not only "system" time but also "REE" time, I suppose `STATS_CMD_GET_TIME` would be a better name

**Review by jforissier — COMMENTED** [review]
> > Why not use a normal TA for this (`TEE_GetSystemTime()` and `TEE_GetREETime()`)?
> 
> Thanks for your comment. We would like to add this cmd in `stats` PTA rather than a user TA because:
> 
>   1. The cmd is to query secure time which is implemented in OS kernel. `TEE_GetSystemTime()` and `TEE_GetREETime()` essentially call into `tee_time_get_sys_time()` and `tee_time_get_ree_time`.
>   2. We hope to test TEE secure without loading a user TA. This can simplifying our test procedure design.
>   3. If we add the cmd into a specific user TA, that means only that TA's client can get TEE time for testing. In our product, the TA is on-demand built into system, so it might not be available in that product. The `stats` PTA looks a common place to add and it will make the cmd available for all products.

OK. In any case it is a very small change, so if that makes your life easier I'm good with it.

I have one minor comment below.

File: core/pta/stats.c (hunk 6 above)
Comment (weizhaojiang) [reply core/pta/stats.c:50]: good point, thanks.

**Comment by jforissier** [comment]
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/pta/stats.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/pta/stats.c:42]: please swap line 40 and 41:
```c
 * (...)
 * uint32_t    Number of failed allocation requests
 * uint32_t    Biggest byte size which allocation failed
 * uint32_t    Biggest byte size which allocation succeeded
 */
#define STATS_CMD_TA_STATS		3

/*
 * STATS_CMD_GET_TIME - Get both REE time and TEE time
 *
 * [out]    value[0].a        REE time as seen by OP-TEE in seconds
 * [out]    value[0].b        REE time as seen by OP-TEE, milliseconds part
 * (...)
```

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` with remaining comment addressed.

**Comment by jforissier** [comment]
`core: pta: stats: implement...` would be a better prefix

**Comment by weizhaojiang** [comment]
> `core: pta: stats: implement...` would be a better prefix

sounds good and done.

**Pull request merged**